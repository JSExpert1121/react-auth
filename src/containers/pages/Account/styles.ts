import { Theme, createStyles } from '@material-ui/core/styles';
import Colors from 'assets/colors';

export const styleFn = (theme: Theme) => createStyles({
    root: {
        width: '100%',
        padding: theme.spacing(3),
        maxWidth: 512,
        minWidth: 480,
    },
    form: {
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
        width: '100%',
        textAlign: 'center'
    },
    logo: {
        padding: theme.spacing(2, 0)
    },
    title: {
        fontSize: '1.5rem',
        color: 'rgba(0, 0, 0, 0.8)',
        padding: theme.spacing(2)
    },
    text: {
        fontSize: '0.875rem',
        color: 'rgba(0, 0, 0, 0.7)',
        padding: theme.spacing(2)
    },
    success: {
        color: 'rgba(0, 0, 0, 0.9)',
        padding: theme.spacing(1)
    },
    error: {
        fontSize: '0.875rem',
        color: Colors.error,
        padding: theme.spacing(0, 0, 0, 2)
    },
    busy: {
        width: '100%',
    },
    submit: {
        margin: theme.spacing(2, 0),
        height: 'calc(100% - 12px)'
    },
    bottomMenu: {
        width: '100%',
        display: 'flex',
        padding: theme.spacing(2, 0)
    }
});