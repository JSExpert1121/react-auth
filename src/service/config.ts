import { oauthClient } from './base';

export const setLanguage = (lang: string) => {
    oauthClient.setLanguage(lang);
}
