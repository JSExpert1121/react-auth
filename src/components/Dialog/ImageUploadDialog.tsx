/**
 * @name ImageUploadDialog
 * @description Dialog for uploading cropped image
 * @version 1.0
 * @todo flexible(for example ratio as props)
 * @author Darko
 */
import * as React from 'react';

// material ui components & icons
import Box from '@material-ui/core/Box';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';
import Button from '@material-ui/core/Button';
import { createStyles, Theme, withStyles, StyledComponentProps } from '@material-ui/core/styles';
import CloudUploadIcon from '@material-ui/icons/CloudUpload';

import ReactCrop from "react-image-crop";
import "react-image-crop/dist/ReactCrop.css";

// component styles
const styles = (theme: Theme) => createStyles({
    root: {
        display: 'flex',
        alignItems: 'center',
    },
    content: {
        textAlign: 'left',
        padding: theme.spacing(1),
        minWidth: 800,
        maxWidth: 960
    },
    preview: {
        display: 'flex',
        justifyContent: 'center',
        height: 240,
        margin: theme.spacing(2, 0)
    },
    avatar: {
        width: 240,
        height: 240,
        backgroundColor: theme.palette.primary.dark,
        color: 'white',
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center'
    },
    fileUpload: {
        display: 'flex',
        flexWrap: 'wrap',
        padding: theme.spacing(1)
    },
    button: {
        borderRadius: 0,
        backgroundColor: theme.palette.primary.light,
        color: '#FFFFFF',
        width: 120,
        margin: theme.spacing(1),
        '&:hover': {
            backgroundColor: theme.palette.primary.dark,
        },
        '&:disabled': {
            backgroundColor: '#FFFFFF',
        },
    }
});

type IUDStatus = 'ready' | 'loaded';

export interface IIUDProps extends StyledComponentProps {
    open: boolean;
    name: string;
    onYes: (url: string) => Promise<any> | void;
    onCancel: () => void;
    url?: string;
    yes?: string;
    ok?: string;
    cancel?: string;
    title?: string;
    message?: string | React.ReactNode;
}

export interface IIUDState {
    busy: boolean;
    photoURL?: string;
    status: IUDStatus;
    crop: { unit: string, height: number, aspect: number };
}

export class ImageUploadDialog extends React.Component<IIUDProps, IIUDState> {

    constructor(props: Readonly<IIUDProps>) {
        super(props);

        this.state = {
            busy: false,
            photoURL: props.url,
            status: 'ready',
            crop: {
                unit: "%",
                height: 100,
                aspect: 1
            }
        }
    }

    imageRef = undefined;
    fileURL = undefined;
    croppedURL = undefined;

    static defaultProps = {
        title: 'Upload Image',
        yes: 'Yes',
        ok: 'OK',
        cancel: 'Cancel'
    };


    handleFileOpened = (e) => {
        const reader = new FileReader();
        reader.addEventListener("load", () => {
            this.setState({
                photoURL: reader.result as string,
                status: 'loaded'
            })
        });
        reader.readAsDataURL(e.target.files[0]);
    }

    onImageLoaded = image => {
        this.imageRef = image;
    }

    getCroppedImg = (image, crop, fileName) => {
        const canvas = document.createElement("canvas");
        const scaleX = image.naturalWidth / image.width;
        const scaleY = image.naturalHeight / image.height;
        canvas.width = crop.width;
        canvas.height = crop.height;
        const ctx = canvas.getContext("2d");

        ctx.drawImage(
            image,
            crop.x * scaleX,
            crop.y * scaleY,
            crop.width * scaleX,
            crop.height * scaleY,
            0,
            0,
            crop.width,
            crop.height
        );

        return new Promise((resolve, reject) => {
            canvas.toBlob(blob => {
                // if (!blob) {
                //     reject(new Error('Canvas is empty'));
                //     return;
                // }
                // blob.name = fileName;
                window.URL.revokeObjectURL(this.fileURL);
                this.fileURL = window.URL.createObjectURL(blob);
                resolve(this.fileURL);
            }, "image/jpeg");
        });
    }

    makeClientCrop = async crop => {
        if (this.imageRef && crop.width && crop.height) {
            this.setState({ busy: true });
            // try {
            const croppedImageUrl = await this.getCroppedImg(
                this.imageRef,
                crop,
                "new_image.jpeg"
            );

            this.croppedURL = croppedImageUrl;
            this.setState({ busy: false });
            // } catch (error) {
            //     console.log('IUD.makeClientCrop: ', error);
            //     this.setState({ busy: false });
            // }
        }
    }

    onCropComplete = async crop => {
        await this.makeClientCrop(crop);
    };

    onCropChange = (crop) => {
        this.setState({ crop });
    };


    render() {
        const { open, onCancel, onYes, title, classes, message, name, cancel, ok } = this.props;
        const { busy, photoURL, crop, status } = this.state;
        return (
            <Dialog
                open={open}
                onClose={onCancel}
                aria-labelledby="image-upload-dialog-title"
                maxWidth='md'
            >
                <DialogTitle id="image-upload-dialog-title">{title}</DialogTitle>
                <DialogContent className={classes.content}>
                    {message && (
                        <DialogContentText id="image-upload-dialog-description">
                            {message}
                        </DialogContentText>
                    )}
                    <Box id='image-upload-dialog-preview' className={classes && classes.preview}>
                        {status === 'ready' && !!photoURL && (
                            <img alt='cropped-avatar' className={classes && classes.avatar} src={photoURL} />
                        )}
                        {status === 'loaded' && !!photoURL && (
                            <ReactCrop
                                src={photoURL}
                                crop={crop}
                                onImageLoaded={this.onImageLoaded}
                                onComplete={this.onCropComplete}
                                onChange={this.onCropChange}
                            />
                        )}
                        {!photoURL && (
                            <Box className={classes && classes.avatar} id='image-upload-dialog-preview-box'>
                                <span style={{ fontSize: 32, color: 'white' }}>
                                    {name.slice(0, 1).toUpperCase()}
                                </span>
                            </Box>
                        )}
                    </Box>
                    <Box className={classes.fileUpload} id='image-upload-dialog-upload'>
                        <input
                            accept="image/*"
                            id="upload-file"
                            type="file"
                            style={{ display: 'none' }}
                            onChange={this.handleFileOpened}
                            disabled={busy}
                        />
                        <label htmlFor="upload-file" style={{ display: 'inline' }}>
                            <Button
                                variant="contained"
                                component="span"
                            >
                                <CloudUploadIcon />
                                &nbsp;&nbsp;Upload a different photo
                                </Button>
                        </label>
                    </Box>
                </DialogContent>
                <DialogActions>
                    {!!onCancel && (
                        <Button id='image-upload-button-cancel' onClick={onCancel} autoFocus className={classes.button}>
                            {cancel}
                        </Button>
                    )}
                    <Button
                        onClick={() => onYes(this.croppedURL)}
                        color="primary"
                        className={classes.button}
                        disabled={busy}
                    >
                        {ok}
                    </Button>
                </DialogActions>
            </Dialog >
        );
    }
}


export default withStyles(styles)(ImageUploadDialog);