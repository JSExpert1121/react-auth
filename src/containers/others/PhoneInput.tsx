/**
 * @name PhoneInput
 * @description verified phone input
 * @version 1.0 (basic)
 * @author Darko
 */
import * as React from 'react';

import Box from '@material-ui/core/Box';
import Link from '@material-ui/core/Link';
import Grid from '@material-ui/core/Grid';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import { makeStyles, Theme } from '@material-ui/core/styles';

// 3rd party libraries
import MuiPhoneInput from 'material-ui-phone-number';
import { MessageDescriptor, injectIntl, IntlShape } from 'react-intl';
import authMsgs from 'translations/messages/page.auth';

// apis & constants
import * as MemberApi from 'service/member';
import { SUPPORTED_COUNTRIES } from 'config/app.config';

const useStyles = makeStyles((theme: Theme) => ({
    root: {
        width: '100%'
    },
    row: {
        paddingTop: theme.spacing(2),
        width: '100%'
    },
    countryCode: {
        marginTop: theme.spacing(1)
    },
    getcode: {
        height: '100%'
    },
}));

interface IPhoneInputProps {
    verify: (country: string, phone: string, flag: boolean) => void;
    setBusy: (flag: boolean) => void;
    intl: IntlShape;
    phoneNo: string;
    countryCode: string;
}

type RecvData = {
    phone: string;
    code: string;
}

export const PhoneInput: React.FC<IPhoneInputProps> = props => {

    const { intl, setBusy, verify } = props;

    const [phone, setPhone] = React.useState(`+${props.countryCode} ${props.phoneNo}`);
    const [country, setCountry] = React.useState(props.countryCode);
    const [verified, setVerified] = React.useState(false);
    const [disabled, setDisabled] = React.useState(false);
    const [code, setCode] = React.useState('');
    const [recvCode, setRecvCode] = React.useState<RecvData[]>([]);
    const [codeMsg, setCodeMsg] = React.useState<MessageDescriptor | undefined>(undefined);

    // phone input event handlers
    const phoneChange = (value: string, data: any) => {
        if (data.dialCode === '82') {
            if (value.length > (data.dialCode.length + 2)) {
                if (value[data.dialCode.length + 2] !== '0') {
                    setPhone(value.slice(0, data.dialCode.length + 2) + '0' + value.slice(data.dialCode.length + 2));
                    setCountry(data.dialCode);
                    return;
                }
            }
        }

        setPhone(value);
        setCountry(data.dialCode);
    }

    const sendCode = async () => {
        // let lang = 'en';
        // if (country === '86') lang = 'zh';
        // else if (country === '82') lang = 'ko';
        // else if (country === '852') lang = 'zh-hk';
        // else if (country === '886') lang = 'zh-tw';
        let realNo = phone.slice(country.length + 2).replace(/[^0-9]+/g, '');   // remove all non-digit characters
        if (country === '82') realNo = realNo.replace(/^0+/g, '');              // for korean phone, remove leading '0's

        setBusy(true);
        try {
            const { response } = await MemberApi.verifyPhone(country, realNo);
            console.log(response);
            if (recvCode.some(item => phone === item.phone)) {
                setRecvCode([...recvCode, { phone, code: response }]);
            } else {
                setRecvCode([{ phone, code: response }]);
            }

            setBusy(false);
            setDisabled(true);
            setCodeMsg(undefined);
        } catch (error) {
            console.log('Signup.SendCode: ', error);
            setBusy(false);
            setCodeMsg(undefined);
        }
    }

    const verifyCode = () => {
        let realNo = phone.slice(country.length + 2).replace(/[^0-9]+/g, '');   // remove all non-digit characters
        if (country === '82') realNo = realNo.replace(/^0+/g, '');              // for korean phone, remove leading '0's
        if (recvCode.some(item => (item.phone === phone && item.code === code))) {
            setVerified(true);
            verify(country, realNo, true);
            setCodeMsg(undefined);
        } else {
            setVerified(false);
            verify(country, realNo, false);
            setCodeMsg(authMsgs.phoneMismatch);
        }
    }

    const tryOtherPhone = () => {
        setVerified(false);
        setDisabled(false);
        setCode('');
    }

    const codeChange = e => {
        setCode(e.target.value);
        setCodeMsg(undefined);
    }

    // phone helper messages
    const notVerifiedCodeHelper = handler => (
        <Box id='notverified-code-helper' style={{ display: 'inline', paddingRight: 8 }}>
            {/* {intl.formatMessage(authMsgs.gotCode)}&nbsp; */}
            <Link onClick={handler} style={{ cursor: 'pointer' }}>
                {intl.formatMessage(authMsgs.phoneResend)}
            </Link>
        </Box>
    )

    // phone helper messages
    const codeHelper = verified ? undefined : notVerifiedCodeHelper(sendCode);
    const sendable = phone.length > (country.length + 2);
    const classes = useStyles({});

    return (
        <Box className={classes.root}>
            <Box className={classes && classes.row}>
                <Grid container spacing={1} className={classes && classes.countryCode}>
                    <Grid item xs={12} sm={8}>
                        <MuiPhoneInput
                            onChange={phoneChange}
                            value={phone}
                            defaultCountry='ko'
                            disableAreaCodes
                            variant='outlined'
                            id='phone-input'
                            fullWidth
                            required
                            label={intl.formatMessage(authMsgs.phoneLabel)}
                            inputProps={{ margin: 'dense', style: { padding: '12px 14px' } }}
                            margin='dense'
                            countryCodeEditable={false}
                            disabled={disabled}
                            onlyCountries={SUPPORTED_COUNTRIES}
                        />
                    </Grid>
                    <Grid item xs={12} sm={4}>
                        <Button
                            fullWidth
                            id='send-code-btn'
                            variant="contained"
                            color="primary"
                            className={classes && classes.getcode}
                            disabled={!sendable || disabled}
                            onClick={sendCode}
                        >
                            {intl.formatMessage(authMsgs.getcode)}
                        </Button>
                    </Grid>
                </Grid>
            </Box>
            {disabled && (
                <Box style={{ width: '100%', paddingTop: 4 }}>
                    {!verified && (
                        <Grid container spacing={1}>
                            <Grid item xs={12} sm={8}>
                                <TextField
                                    id='phone-code'
                                    name='phoneCode'
                                    variant='outlined'
                                    label={intl.formatMessage(authMsgs.phoneCode)}
                                    fullWidth
                                    margin='dense'
                                    error={!!codeMsg}
                                    helperText={!!codeMsg && intl.formatMessage(codeMsg)}
                                    value={code}
                                    onChange={codeChange}
                                />
                            </Grid>
                            <Grid item xs={12} sm={4} style={{ paddingTop: 12, paddingBottom: 8 }}>
                                <Button
                                    fullWidth
                                    variant="contained"
                                    color="primary"
                                    id="verify-btn"
                                    className={classes && classes.getcode}
                                    disabled={code.length !== 6}
                                    onClick={verifyCode}
                                >
                                    {intl.formatMessage(authMsgs.phoneVerify)}
                                </Button>
                            </Grid>
                        </Grid>
                    )}
                    {codeHelper}
                    <Link id='try-otherphone-link' onClick={tryOtherPhone} style={{ cursor: 'pointer' }}>
                        {intl.formatMessage(authMsgs.phoneTryOther)}
                    </Link>{''}
                </Box>
            )}
        </Box>
    )
}

export default injectIntl(PhoneInput);