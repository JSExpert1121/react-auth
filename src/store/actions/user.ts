/**
 * @name GlobalAction
 * @description user management. sign up, sign in, reset password, ...
 * @version 0.1 (only skeleton)
 * @version 0.5 (signup)
 * @version 1.0 (signup/login/logout)
 * @todo according to spec
 * @author Darko
 */
import { User, UserProfile } from 'types/user';
import * as MemberApi from 'service/member';
import { Gender } from 'types/global';

import {
    USER_LOGIN_SUCCESS,
    USER_GET_USER_SUCCESS,
    USER_UPDATE_USER_SUCCESS,
    USER_GET_PROFILE_SUCCESS
} from '../constants';

export const loginSuccess = (token: string) => ({
    type: USER_LOGIN_SUCCESS,
    payload: token
});

export const getuserSuccess = (user: User) => ({
    type: USER_GET_USER_SUCCESS,
    payload: user
});

export const getprofileSuccess = (data: UserProfile) => ({
    type: USER_GET_PROFILE_SUCCESS,
    payload: data
});

export const updateUserSuccess = (user: User) => ({
    type: USER_UPDATE_USER_SUCCESS,
    payload: user
});

const clearLogin = dispatch => {
    dispatch(loginSuccess(undefined));
    dispatch(getuserSuccess(undefined));
    dispatch(getprofileSuccess(undefined));
}

let timeHandler = undefined;
export const refreshToken = () => async dispatch => {
    const newToken = await MemberApi.refresh();
    dispatch(loginSuccess(newToken));
    if (newToken) {
        timeHandler = setTimeout(() => dispatch(refreshToken()), 3500000);
    }
}

export const signup = (email: string, pass: string, name: string, code: string, phone: string, agrees: Array<any>) => async dispatch => {

    let data = undefined;
    data = await MemberApi.register(email, pass, name);
    // console.log(data);
    const token = await MemberApi.login(email, pass);
    dispatch(loginSuccess(token));

    data = await MemberApi.agreeTAC(token, agrees);
    data = await MemberApi.updateProfile(token, code, phone, true);

    // get user data
    data = await MemberApi.getUser(token);
    dispatch(getuserSuccess(data));

    // get user profile
    data = await MemberApi.getProfile(token);
    dispatch(getprofileSuccess(data));
}

export const login = (email: string, pass: string) => async dispatch => {
    const token = await MemberApi.login(email, pass);
    dispatch(loginSuccess(token));

    // get user data
    let data = await MemberApi.getUser(token);
    dispatch(getuserSuccess(data));

    // get user profile
    data = await MemberApi.getProfile(token);
    dispatch(getprofileSuccess(data));
}

export const logout = () => dispatch => MemberApi.logout().then(data => {
    clearLogin(dispatch);
    timeHandler && clearTimeout(timeHandler);
    timeHandler = undefined;
});

export const getUser = (token: string, id?: number) => dispatch => MemberApi.getUser(token, id).then(data => {
    dispatch(getuserSuccess(data));
    return data;
});

export const updateUser = (token: string, newMail: string, password: string, name: string, id?: number) => dispatch => MemberApi.updateUser(token, newMail, password, name, id).then(data => {
    dispatch(updateUserSuccess(data));
    return data;
});

export const deleteUser = (token: string, id?: number) => dispatch => MemberApi.deleteUser(token, id).then(data => {
    clearLogin(dispatch);
});

export const getProfile = (token: string, id?: number) => dispatch => MemberApi.getProfile(token, id).then(data => {
    dispatch(getprofileSuccess(data));
    return data;
});

export const updateProfile = (
    token: string,
    country: string,
    phone: string,
    phoneVerified: boolean,
    id?: number,
    birth?: string,
    gender?: Gender,
    photo?: string,
) => dispatch => MemberApi.updateProfile(token, country, phone, phoneVerified, id, birth, gender, photo).then(data => {
    dispatch(getprofileSuccess(data));
    return data;
});
