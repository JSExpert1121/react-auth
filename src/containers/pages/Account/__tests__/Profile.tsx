import React from 'react';
import { MemoryRouter, withRouter } from 'react-router-dom';

import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16/build';

import { createShallow, createMount } from '@material-ui/core/test-utils';
import { ThemeProvider } from '@material-ui/styles';
import { IntlProvider, injectIntl } from "react-intl";
import { act } from 'react-dom/test-utils';

import Paper from '@material-ui/core/Paper';
import FormControl from '@material-ui/core/FormControl';
import Avatar from '@material-ui/core/Avatar';
import Box from '@material-ui/core/Box';
import IconButton from '@material-ui/core/IconButton';
import Button from '@material-ui/core/Button';
import InputLabel from '@material-ui/core/InputLabel';
import Select from '@material-ui/core/Select';
import Grid from '@material-ui/core/Grid';
import LinearProgress from '@material-ui/core/LinearProgress';

import {
    DatePicker,
} from '@material-ui/pickers';

import PhoneInput from 'containers/others/PhoneInput';
import ImageUploadDialog from 'components/Dialog/ImageUploadDialog';

import { translationMessages } from 'config/i18n.config';
import messages from 'translations/messages/page.account';
import { User, UserProfile } from 'types/user';

// store mock
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

import DefProfilePage, { ProfilePage, StyledWarning } from '../Profile';
import theme from 'config/theme.config';
jest.mock('service/member');

configure({ adapter: new Adapter() });
const ConnectedProfilePage = withRouter(DefProfilePage);


describe('containers/pages/Account/Profile', () => {

    const shallow = createShallow();
    const user: User = { id: 100, name: 'testuser', username: 'test@test.com', enabled: true };
    const priv_token = 'test@test.com__private_token';
    const profile: UserProfile = {
        cellphoneCountryCode: '82',
        cellphone: '1000000000',
        emailVerified: false,
        cellphoneVerified: true
    }
    const mock_photo = 'https://google.com/image.png';
    const fnIntl = { formatMessage: jest.fn() };
    fnIntl.formatMessage.mockReturnValue('My Profile');

    const setTitle = jest.fn();
    const updateProfile = jest.fn();
    updateProfile.mockResolvedValueOnce(true).mockRejectedValueOnce(false);
    const showMessage = jest.fn();

    const wrapper = shallow(
        <ProfilePage
            intl={fnIntl}
            profile={profile}
            token={priv_token}
            user={user}
            setTitle={setTitle}
            updateProfile={updateProfile}
            classes={{}}
            showMessage={showMessage}
        />
    );

    test('Snapshot test', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('SetTitle', done => {
        wrapper.update();
        expect(fnIntl.formatMessage).toHaveBeenCalledWith(messages.profileTitle);
        expect(setTitle).toHaveBeenCalledWith('My Profile');
        done();
    });

    test('UI components', () => {
        expect(wrapper.find(Paper).find(FormControl).first().find(Avatar)).toHaveLength(1);
        expect(wrapper.find(Paper).find(FormControl).first().find(Avatar).find(IconButton)).toHaveLength(0);
        expect(wrapper.find(Paper).find(FormControl).find(Box).first().find(LinearProgress)).toHaveLength(1);
        expect(wrapper.find(Paper).find(FormControl).find(Box).first().find(PhoneInput)).toHaveLength(1);
        expect(wrapper.find(Paper).find(FormControl).find(Box).first().find(Grid).at(1).find(FormControl)).toHaveLength(1);
        expect(wrapper.find(Paper).find(FormControl).find(Box).first().find(Grid).at(1).find(FormControl).find(InputLabel)).toHaveLength(1);
        expect(wrapper.find(ImageUploadDialog).props().open).toBe(false);
        expect(wrapper.find(ImageUploadDialog).props().onYes).toBeInstanceOf(Function);
        expect(wrapper.find(ImageUploadDialog).props().onCancel).toBeInstanceOf(Function);
    });

    test('User interaction - enter avatar', done => {
        act(() => {
            wrapper.find(Avatar).simulate('mouseenter');
        });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(Paper).find(FormControl).first().find(Avatar).find(IconButton)).toHaveLength(1);
            done();
        })
    });

    test('User interaction - edit avatar', done => {
        act(() => {
            wrapper.find(Paper).find(FormControl).first().find(Avatar).find(IconButton).simulate('click');
        });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(ImageUploadDialog).props().open).toBe(true);
            expect(wrapper.find(ImageUploadDialog).props().name).toEqual(user.name);
            done();
        })
    });

    test('User interaction - close avatar edit', done => {
        wrapper.find(ImageUploadDialog).props().onCancel();

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(ImageUploadDialog).props().open).toBe(false);
            done();
        })
    });

    test('User interaction - reopen edit avatar', done => {
        act(() => {
            wrapper.find(Paper).find(FormControl).first().find(Avatar).find(IconButton).simulate('click');
        });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(ImageUploadDialog).props().open).toBe(true);
            done();
        })
    });

    test('User interaction - save avatar', done => {
        wrapper.find(ImageUploadDialog).props().onYes(mock_photo);

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(ImageUploadDialog).props().open).toBe(false);
            done();
        })
    });

    test('User interaction - PhoneInput completion handler', done => {
        wrapper.find(PhoneInput).props().verify('82', '10000', true);

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(PhoneInput).props().phoneNo).toEqual('10000');
            done();
        });
    });

    test('User interaction - birthday change - 1', done => {
        wrapper.find(DatePicker).props().onChange(new Date(1994, 10, 8));
        // dateBirth: '1994-11-08'

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(DatePicker).props().value).toEqual('1994-11-08');
            done();
        })
    });

    test('User interaction - birthday change - 2', done => {
        wrapper.find(DatePicker).props().onChange(new Date(1994, 0, 28));
        // dateBirth: '1994-01-28'

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(DatePicker).props().value).toEqual('1994-01-28');
            done();
        })
    });

    test('User interaction - gender change', done => {
        wrapper.find(Select).props().onChange({ target: { value: 'FEMALE' } });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(Select).props().value).toEqual('FEMALE');
            done();
        })
    });

    test('User interaction - enter avatar', done => {
        act(() => {
            wrapper.find(Avatar).simulate('mouseenter');
        });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(Paper).find(FormControl).first().find(Avatar).find(IconButton)).toHaveLength(1);
            done();
        })
    });

    test('User interaction - leave avatar', done => {
        act(() => {
            wrapper.find(Avatar).simulate('mouseleave');
        });

        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(Paper).find(FormControl).first().find(Avatar).find(IconButton)).toHaveLength(0);
            done();
        })
    });

    test('User interaction - save - success', async done => {
        await wrapper.find(Button).props().onClick(undefined);

        setImmediate(() => {
            wrapper.update();
            expect(updateProfile).toHaveBeenCalledWith(
                priv_token, '82', '10000', true, undefined,
                '1994-01-28', 'FEMALE', mock_photo
            );
            expect(fnIntl.formatMessage).toHaveBeenCalledWith(messages.profileSuccess);
            expect(showMessage).toHaveBeenCalledWith(true, 'My Profile');
            done();
        })
    });

    test('User interaction - save - error', async done => {
        await wrapper.find(Button).props().onClick(undefined);

        setImmediate(() => {
            wrapper.update();
            expect(updateProfile).toHaveBeenCalledWith(
                priv_token, '82', '10000', true, undefined,
                '1994-01-28', 'FEMALE', mock_photo
            );
            expect(fnIntl.formatMessage).toHaveBeenCalledWith(messages.profileError);
            expect(showMessage).toHaveBeenCalledWith(false, 'My Profile');
            done();
        })
    });

    test('User interaction - phoneinput busy', done => {
        wrapper.find(PhoneInput).props().setBusy(true);
        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(LinearProgress).props().style).toEqual({
                visibility: 'visible'
            });
            done();
        })
    });

    test('User interaction - phoneinput not busy', done => {
        wrapper.find(PhoneInput).props().setBusy(false);
        setImmediate(() => {
            wrapper.update();
            expect(wrapper.find(LinearProgress).props().style).toEqual({
                visibility: 'hidden'
            });
            done();
        })
    });
});

describe('containers/pages/Account/Profile - variable props', () => {

    const shallow = createShallow();

    test('UI components without photo', () => {
        const user: User = { id: 100, name: 'testuser', username: 'test@test.com', enabled: true };
        const priv_token = 'test@test.com__private_token';
        const profile: UserProfile = {
            cellphoneCountryCode: '82',
            cellphone: '1000000000',
            emailVerified: false,
            cellphoneVerified: true
        }

        const fnIntl = { formatMessage: jest.fn() };
        const setTitle = jest.fn();
        const updateProfile = jest.fn();

        const wrapper = shallow(
            <ProfilePage
                intl={fnIntl}
                profile={profile}
                token={priv_token}
                user={user}
                setTitle={setTitle}
                updateProfile={updateProfile}
                classes={{}}
            />
        );

        expect(wrapper.find(Paper).find(FormControl).first().find(Avatar).props().src).toBe(undefined);
        expect(wrapper.find(Paper).find(FormControl).first().find(Avatar).find(IconButton)).toHaveLength(0);
    });

    test('UI components with photo', () => {
        const user: User = { id: 100, name: 'testuser', username: 'test@test.com', enabled: true };
        const priv_token = 'test@test.com__private_token';
        const mock_photo = 'https://google.com/image.png';
        const profile: UserProfile = {
            cellphoneCountryCode: '82',
            cellphone: '1000000000',
            emailVerified: false,
            cellphoneVerified: true,
            photoUrl: mock_photo
        }

        const fnIntl = { formatMessage: jest.fn() };
        const setTitle = jest.fn();
        const updateProfile = jest.fn();

        const wrapper = shallow(
            <ProfilePage
                intl={fnIntl}
                profile={profile}
                token={priv_token}
                user={user}
                setTitle={setTitle}
                updateProfile={updateProfile}
                classes={{}}
            />
        );

        expect(wrapper.find(Paper).find(FormControl).first().find(Avatar).props().src).toEqual(mock_photo);
    });

    test('UI components not verified email, phone', () => {
        const user: User = { id: 100, name: 'testuser', username: 'test@test.com', enabled: true };
        const priv_token = 'test@test.com__private_token';
        const mock_photo = 'https://google.com/image.png';
        const profile: UserProfile = {
            cellphoneCountryCode: '82',
            cellphone: '1000000000',
            emailVerified: false,
            cellphoneVerified: false,
            photoUrl: mock_photo
        }

        const fnIntl = { formatMessage: jest.fn() };
        const setTitle = jest.fn();
        const updateProfile = jest.fn();

        const wrapper = shallow(
            <ProfilePage
                intl={fnIntl}
                profile={profile}
                token={priv_token}
                user={user}
                setTitle={setTitle}
                updateProfile={updateProfile}
                classes={{}}
            />
        );

        expect(wrapper.find(Paper).find(StyledWarning)).toHaveLength(2);
    });

    test('UI components verified email, phone', () => {
        const user: User = { id: 100, name: 'testuser', username: 'test@test.com', enabled: true };
        const priv_token = 'test@test.com__private_token';
        const mock_photo = 'https://google.com/image.png';
        const profile: UserProfile = {
            cellphoneCountryCode: '82',
            cellphone: '1000000000',
            emailVerified: true,
            cellphoneVerified: true,
            photoUrl: mock_photo
        }

        const fnIntl = { formatMessage: jest.fn() };
        const setTitle = jest.fn();
        const updateProfile = jest.fn();

        const wrapper = shallow(
            <ProfilePage
                intl={fnIntl}
                profile={profile}
                token={priv_token}
                user={user}
                setTitle={setTitle}
                updateProfile={updateProfile}
                classes={{}}
            />
        );

        expect(wrapper.find(Paper).find(StyledWarning)).toHaveLength(0);
        expect(wrapper.find(PhoneInput).props().phoneNo).toEqual('1000000000');
        expect(wrapper.find(PhoneInput).props().countryCode).toEqual('82');
    });

    test('UI components without phone', () => {
        const user: User = { id: 100, name: 'testuser', username: 'test@test.com', enabled: true };
        const priv_token = 'test@test.com__private_token';
        const mock_photo = 'https://google.com/image.png';
        const profile: UserProfile = {
            cellphoneCountryCode: undefined,
            cellphone: undefined,
            emailVerified: true,
            cellphoneVerified: true,
            photoUrl: mock_photo
        }

        const fnIntl = { formatMessage: jest.fn() };
        const setTitle = jest.fn();
        const updateProfile = jest.fn();

        const wrapper = shallow(
            <ProfilePage
                intl={fnIntl}
                profile={profile}
                token={priv_token}
                user={user}
                setTitle={setTitle}
                updateProfile={updateProfile}
                classes={{}}
            />
        );

        expect(wrapper.find(Paper).find(StyledWarning)).toHaveLength(0);
        expect(wrapper.find(PhoneInput).props().phoneNo).toEqual('');
        expect(wrapper.find(PhoneInput).props().countryCode).toEqual('');
    });

    test('StyledWarning', () => {
        const shallow = createShallow();
        const wrapper = shallow(<StyledWarning />);
    })

    test('real mount', () => {
        const mount = createMount();

        const user: User = { id: 100, name: 'testuser', username: 'test@test.com', enabled: true };
        const profile: UserProfile = {
            cellphoneCountryCode: undefined,
            cellphone: undefined,
            emailVerified: true,
            cellphoneVerified: true,
            photoUrl: undefined
        }
        const store = mockStore({
            user: { user, profile, token: undefined },
        });

        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <IntlProvider locale='en' messages={translationMessages.en}>
                        <ThemeProvider theme={theme}>
                            {<ConnectedProfilePage />}
                        </ThemeProvider>
                    </IntlProvider>
                </MemoryRouter>
            </Provider>
        );
    });
});

