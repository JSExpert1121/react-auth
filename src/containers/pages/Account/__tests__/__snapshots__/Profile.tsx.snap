// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/pages/Account/Profile Snapshot test 1`] = `ShallowWrapper {}`;
